File: src/com/trinitcore/sql/SQL.java
package com.trinitcore.sql;

import com.trinitcore.sql.queryObjects.QueryObject;
import com.trinitcore.sql.queryObjects.returnableQueries.Select;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by cormacpjkinsella on 10/9/16.
 */
public class SQL {

    public String query;
    public List<Object> parameters = new ArrayList<Object>();
    public QueryObject queryObject;

    public ResultSet resultSet = null;
    public PreparedStatement preparedStmt = null;
    public static Configuration configuration;
    public Connection conn = null;

    public static void setConfiguration(Configuration extendedConfiguration) {
        configuration = extendedConfiguration;
    }

    public SQL() {

    }

    public SQL(String query, List<Object> parameters) {
        this.query = query;
        this.parameters = parameters;
    }

    public SQL(Select selectQuery) {
        this.query = selectQuery.query;
        this.queryObject = selectQuery;

        System.out.println(query);
    }

    public SQL query(boolean returnable) {
        try {
            conn = configuration.getConnection();
            this.preparedStmt = conn.prepareStatement(query);
            int count = 1;
            for (Object value : this.parameters){
                this.preparedStmt.setObject(count,value);
                count++;
            }
            if (returnable){
                this.resultSet = this.preparedStmt.executeQuery();
                System.out.println("QUERY OUT: "+this.resultSet.getStatement().toString());
            }
            else this.preparedStmt.execute();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return this;
    }

    public SQL close() {
        try {
            if (conn == null) return this;
            else if (conn.isClosed()) return this;
            this.conn.close();
            this.preparedStmt.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return this;
    }


}

File: src/com/trinitcore/sqlv2/queryObjects/SQL.kt
package com.trinitcore.sqlv2.queryObjects

import com.trinitcore.sqlv2.commonUtils.Defaults
import com.trinitcore.sqlv2.commonUtils.NullValue
import com.trinitcore.sqlv2.commonUtils.then
import com.trinitcore.sqlv2.queryUtils.connection.ConnectionManager
import org.postgresql.util.PSQLException
import java.sql.PreparedStatement
import java.sql.ResultSet

/**
 * Created by Cormac on 18/08/2017.
 */

object SQL {

    public lateinit var sharedConnection: ConnectionManager

    public fun returnable(query: String, parameters: Array<Any>, returnColumnKey: Boolean = false): ResultSet? {
            val updateParams = arrayOf(Defaults.indexColumnKey)
            var statement: PreparedStatement? = null
            statement = getStatement(query = query, parameters = parameters,
                    resultSetParameters = returnColumnKey then updateParams ?: emptyArray()
            )

            //try {
                if (returnColumnKey) {
                    statement.executeUpdate()
                    return statement.generatedKeys
                }
                return statement.executeQuery()
            /*} catch (e: PSQLException) {
                println("Reopening connection")
                return SQL.session {
                    return returnable(query, parameters, returnColumnKey)
                } as ResultSet?
            }*/

    }

    public fun noneReturnable(query: String, parameters: Array<Any> = emptyArray()): Boolean {
            return !(getStatement(query, parameters).execute() ?: true)
    }

    private fun getStatement(query: String, parameters: Array<Any>, resultSetParameters: Array<String> = emptyArray()): PreparedStatement {
        val finalparameters = parameters.map { if (it is NullValue) null else it }

        /*
        if (sharedConnection.currentConnection == null || sharedConnection.currentConnection?.isClosed == true) {
            return SQL.session {
                return getStatement(query, parameters, resultSetParameters)
            } as PreparedStatement
        }
        */
        val statement = sharedConnection.currentConnection!!.prepareStatement(query, resultSetParameters)

        var count = 0
        for (p in finalparameters) statement.setObject(++count, p)

        return statement
    }

    public inline fun session(stream: () -> Unit): Any? {
        var shouldCloseConnection = true
        shouldCloseConnection = sharedConnection.open()

        val s = stream()
        // if (shouldCloseConnection) {
            // sharedConnection.close()
        // }
        return s
    }
}